crates/containerd-shim-wasmedge/src/instance.rs

```diff
@@ -103,171 +134,158 @@ pub fn reset_stdio() {
             exit_code: Arc::new((Mutex::new(None), Condvar::new())),
-            engine: cfg.get_engine(),
             stdin: cfg.get_stdin().unwrap_or_default(),
             stdout: cfg.get_stdout().unwrap_or_default(),
             stderr: cfg.get_stderr().unwrap_or_default(),
-            bundle: cfg.get_bundle().unwrap_or_default(),
-            pidfd: Arc::new(Mutex::new(None)),
+            bundle,
         }
     }
-    fn start(&self) -> Result<u32, Error> {
-        let engine = self.engine.clone();
-        let stdin = self.stdin.clone();
-        let stdout = self.stdout.clone();
-        let stderr = self.stderr.clone();
 
+    fn start(&self) -> Result<u32, Error> {
         debug!("preparing module");
-        let spec = load_spec(self.bundle.clone())?;
-
-        let vm = prepare_module(engine, &spec, stdin, stdout, stderr)
-            .map_err(|e| Error::Others(format!("error setting up module: {}", e)))?;
-
-        let cg = oci::get_cgroup(&spec)?;
-
-        oci::setup_cgroup(cg.as_ref(), &spec)
-            .map_err(|e| Error::Others(format!("error setting up cgroups: {}", e)))?;
-        let res = unsafe { exec::fork(Some(cg.as_ref())) }?;
-        match res {
-            exec::Context::Parent(tid, pidfd) => {
-                let mut lr = self.pidfd.lock().unwrap();
-                *lr = Some(pidfd.clone());
-
-                debug!("started wasi instance with tid {}", tid);
-
-                let code = self.exit_code.clone();
-
-                let _ = thread::spawn(move || {
-                    let (lock, cvar) = &*code;
-                    let status = match pidfd.wait() {
-                        Ok(status) => status,
-                        Err(e) => {
-                            error!("error waiting for pid {}: {}", tid, e);
-                            cvar.notify_all();
-                            return;
-                        }
-                    };
-
-                    debug!("wasi instance exited with status {}", status.status);
-                    let mut ec = lock.lock().unwrap();
-                    *ec = Some((status.status, Utc::now()));
-                    drop(ec);
-                    cvar.notify_all();
-                });
-                Ok(tid)
-            }
-            exec::Context::Child => {
-                // child process
-
-                // TODO: How to get exit code?
-                // This was relatively straight forward in go, but wasi and wasmtime are totally separate things in rust.
-                let _ret = match vm.run_func(Some("main"), "_start", params!()) {
-                    Ok(_) => std::process::exit(0),
-                    Err(_) => std::process::exit(137),
-                };
-            }
-        }
+
+        let syscall = create_syscall();
+        let mut container = ContainerBuilder::new(self.id.clone(), syscall.as_ref())
+            .with_executor(vec![Box::new(WasmEdgeExecutor {
```

Comment:
It can be done; youki actually does. 
https://github.com/containers/youki/blob/d1e19fdbe3229999c287949e31615b4251c90d17/crates/youki/src/workload/executor.rs#L3-L13

At an abstract level, you can implement what is to be executed in the container.

Reply:
Thanks for the link! It looks like three out of 4 executors are wasm executors. I am assuming the default executor is the one that can execute normal container workloads, right?

Reply:
Yeah, the `DefaultExecutor` is the one that executes the container's process https://github.com/containers/youki/blob/d1e19fdbe3229999c287949e31615b4251c90d17/crates/libcontainer/src/workload/default.rs#L14